daemon/create.go

```diff
@@ -92,17 +96,65 @@ func (daemon *Daemon) Create(config *runconfig.Config, hostConfig *runconfig.Hos
 	if err := daemon.createRootfs(container); err != nil {
 		return nil, nil, err
 	}
-	if hostConfig != nil {
-		if err := daemon.setHostConfig(container, hostConfig); err != nil {
-			return nil, nil, err
-		}
+	if err := daemon.setHostConfig(container, hostConfig); err != nil {
+		return nil, nil, err
 	}
 	if err := container.Mount(); err != nil {
 		return nil, nil, err
 	}
 	defer container.Unmount()
-	if err := container.prepareVolumes(); err != nil {
-		return nil, nil, err
+
+	uniqueDestination := func(destination string) bool {
+		for _, volumeConfig := range container.VolumeConfig {
+			if volumeConfig.Destination == destination {
+				return false
+			}
+		}
+		return true
+	}
+
+	for spec := range config.Volumes {
+		var (
+			name, destination string
+			parts             = strings.Split(spec, ":")
+		)
+		switch len(parts) {
+		case 2:
+			name, destination = parts[0], filepath.Clean(parts[1])
+		default:
+			name = stringid.GenerateRandomID()
+			destination = filepath.Clean(parts[0])
+		}
+		// Skip volumes for which we already have something mounted on that
+		// destination because of a --volume-from.
+		if !uniqueDestination(destination) {
+			continue
+		}
+		path, err := container.GetResourcePath(destination)
+		if err != nil {
+			return nil, nil, err
+		}
+		if stat, err := os.Stat(path); err == nil && !stat.IsDir() {
+			return nil, nil, fmt.Errorf("cannot mount volume over existing file, file exists %s", path)
+		}
+		v, err := daemon.volumeDriver.Create(name)
+		if err != nil {
+			return nil, nil, err
+		}
+		rootfs, err := symlink.FollowSymlinkInScope(filepath.Join(container.basefs, destination), container.basefs)
+		if err != nil {
+			return nil, nil, err
+		}
+		if path, err = v.Mount(); err != nil {
+			return nil, nil, err
+		}
+		copyExistingContents(rootfs, path)
+		container.volumes = append(container.volumes, v)
+		container.VolumeConfig[v.Name()] = &VolumeConfig{
+			Plugin:      daemon.volumeDriver.Name(),
```

Comment:
Here's the thing. A plugin provides extensions of various types. It can provide a volume driver, a network driver, but maybe non-drivers, like maybe an event listener. These types are based on the interfaces they implement. A plugin is not just a driver or listener, it could be any. So the concept of plugin is just a thing that provides extensions in the form of implemented interfaces.

In this case, you're not even talking about a plugin, you're just talking about a driver. What driver does this volume use? Therefore, we call it Driver.